components/scream/src/share/grid/abstract_grid.cpp

```diff
@@ -249,6 +249,204 @@ void AbstractGrid::reset_num_vertical_lev (const int num_vertical_lev) {
   //       invalidate all geo data whose FieldLayout contains LEV/ILEV
 }
 
+auto
+AbstractGrid::get_unique_gids () const
+ -> dofs_list_type
+{
+  // Gather local sizes across all ranks
+  std::vector<int> ngids (m_comm.size());
+  ngids[m_comm.rank()] = get_num_local_dofs();
+  m_comm.all_gather(ngids.data(),1);
+
+  std::vector<int> offsets (m_comm.size()+1,0);
+  for (int pid=1; pid<=m_comm.size(); ++pid) {
+    offsets[pid] = offsets[pid-1] + ngids[pid-1];
+  }
+  EKAT_REQUIRE_MSG (offsets[m_comm.size()]==m_num_global_dofs,
+      "Error! Something went wrong while computing offsets in AbstractGrid::get_unique_grid.\n");
+
+  // Gather all dofs
+  const auto mpi_gid_t = ekat::get_mpi_type<gid_type>();
+  std::vector<gid_type> all_gids (m_num_global_dofs);
+  MPI_Allgatherv (m_dofs_gids_host.data(),m_num_local_dofs,mpi_gid_t,
+                  all_gids.data(),ngids.data(),offsets.data(),
+                  mpi_gid_t,m_comm.mpi_comm());
+
+  // Figure out unique dofs
+  std::vector<gid_type> unique_dofs;
+  const auto all_gids_beg = all_gids.begin();
+  const auto all_gids_end = all_gids.begin() + offsets[m_comm.rank()];
+  const auto my_gids_beg = m_dofs_gids_host.data();
+  const auto my_gids_end = m_dofs_gids_host.data() + m_num_local_dofs;
+  for (auto it=my_gids_beg; it!=my_gids_end; ++it) {
+    if (std::find(all_gids_beg,all_gids_end,*it)==all_gids_end) {
+      unique_dofs.push_back(*it);
+    }
+  }
+
+  dofs_list_type unique_gids_d("",unique_dofs.size());
+  auto unique_gids_h = Kokkos::create_mirror_view(unique_gids_d);
+  std::memcpy(unique_gids_h.data(),unique_dofs.data(),sizeof(gid_type)*unique_dofs.size());
+  Kokkos::deep_copy(unique_gids_d,unique_gids_h);
+  return unique_gids_h;
+}
+
+auto AbstractGrid::
```

Comment:
It actually does have an explanation, in the hpp file. I tend to put a comment explaining what the fcn does in the hpp, since that's where the API is.